spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.zookeeper.metadata;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import com.gs.collections.impl.map.mutable.ConcurrentHashMap;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.ChildData;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.CloseableUtils;
+import org.apache.curator.utils.EnsurePath;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+
+/**
+ * Zookeeper-based {@link ListenableMetadataStore} based on a Zookeeper node. Values are stored in the children node,
+ * the names of which are stored as keys.
+ *
+ * @author Marius Bogoevici
+ */
+public class ZookeeperMetadataStore implements ListenableMetadataStore, SmartLifecycle {
+
+	private final Object lifecycleMonitor = new Object();
+
+	private final CuratorFramework client;
+
+	private final List<MetadataStoreListener> listeners = new CopyOnWriteArrayList<>();
+
+	/**
+	 * An internal map storing local updates, ensuring that they have precedence if the cache contains stale data.
+	 * As changes are propagated back from Zookeeper to the cache, entries are removed.
+	 */
+	private final ConcurrentMap<String, LocalChildData> updateMap = new ConcurrentHashMap<>();
+
+	private volatile String root = "/SpringIntegration-MetadataStore";
+
+	private volatile String encoding = "UTF-8";
+
+	private volatile PathChildrenCache cache;
+
+	private volatile boolean running = false;
+
+	private volatile boolean autoStartup = false;
+
+	private volatile int phase = Integer.MAX_VALUE;
+
+	public ZookeeperMetadataStore(CuratorFramework client) throws Exception {
+		Assert.notNull(client, "Client cannot be null");
+		this.client = client;
+	}
+
+	/**
+	 * Encoding to use when storing data in ZooKeeper
+	 *
+	 * @param encoding encoding as text
+	 */
+	public void setEncoding(String encoding) {
+		Assert.notNull(encoding, "'encoding' must not be null.");
+		this.encoding = encoding;
+	}
+
+	/**
+	 * Root node - store entries are children of this node
+	 *
+	 * @param root encoding as text
+	 */
+	public void setRoot(String root) {
+		Assert.notNull(root, "'root' must not be null.");
+		Assert.isTrue(root.startsWith("/"), "'root' must start with '/'");
+		Assert.isTrue(!root.endsWith("/"), "'root' must not end with '/'");
```

Comment:
I think we want the starting "/" because it's pretty much standard for ZK configurations. We can be lenient, but this is still ZK-specific. There's no notion of relative paths at this point. My 2 cents. The trailing '/' can be removed transparently.